packages/lesswrong/components/users/DialogueMatchingPage.tsx

```diff
@@ -353,41 +653,246 @@ const UserTopTags = ({ classes, targetUserId }: { classes: ClassesType, targetUs
+  const [formatNotes, setFormatNotes] = useState(dialogueCheck.matchPreference?.formatNotes || "");
+
+  const { create, called, loading: loadingCreatedMatchPreference, data: newMatchPreference } = useCreate({
+    collectionName: "DialogueMatchPreferences",
+    fragmentName: "DialogueMatchPreferencesDefaultFragment",
+  })
+
+  const navigate = useNavigate();
+
+  const onSubmit = async () => {
+    const response = await create({
+      data: {
+        dialogueCheckId: dialogueCheckId,
+        topicPreferences: topicPreferences.map(topic => ({...topic, matchedPersonPreference: undefined, preference: topic.preference || "No"})),
+        topicNotes: topicNotes,
+        syncPreference: formatSync,
+        asyncPreference: formatAsync,
+        formatNotes: formatNotes,
+      }
+    })
+
+    const redirectId = response.data?.createDialogueMatchPreference.data.generatedDialogueId
+    redirect(redirectId, navigate)
+  }
+
+  const { loading, error, data } = useQuery(gql`
+    query getTopicRecommendations($userId: String!, $targetUserId: String!, $limit: Int!) {
+      GetTwoUserTopicRecommendations(userId: $userId, targetUserId: $targetUserId, limit: $limit) {
+        _id
+        contents {
+          html
+          plaintextMainText
+        }
+      }
+    }
+  `, {
+    variables: { userId, targetUserId, limit:4 },
+  });
+
+
+  const topicRecommendations: CommentsList[] = data?.GetTwoUserTopicRecommendations; // Note CommentsList is too permissive here, but making my own type seemed too hard
+
+  
+  const ownTopicDict = Object.fromEntries(dialogueCheck.matchPreference?.topicPreferences?.filter(topic => topic.preference === "Yes").map(topic => [topic.text, topic]) || [])
+  const matchedPersonTopicDict = Object.fromEntries(dialogueCheck.matchingMatchPreference?.topicPreferences?.filter(topic => topic.preference === "Yes").map(topic => [topic.text, {...topic, preference: undefined, matchedPersonPreference: topic.preference}]) || [])
+  const mergedTopicDict = mergeWith(ownTopicDict, matchedPersonTopicDict, (ownTopic, matchedPersonTopic) => ({...matchedPersonTopic, ...ownTopic}))
+  const [topicPreferences, setTopicPreferences] = useState<ExtendedDialogueMatchPreferenceTopic[]>(Object.values(mergedTopicDict))
+
+  useEffect(() => setTopicPreferences(topicPreferences => {
+    const existingTopicDict = Object.fromEntries(topicPreferences.map(topic => [topic.text, topic]))
+    const newRecommendedTopicDict = Object.fromEntries(topicRecommendations?.map(comment => [comment.contents?.plaintextMainText || '', {
+      text: comment.contents?.plaintextMainText || '',
+      preference: 'No' as const,
+      commentSourceId: comment._id
+    }]) || [])
+    const mergedTopicDict = mergeWith(existingTopicDict, newRecommendedTopicDict, (existingTopic, newRecommendedTopic) => ({...newRecommendedTopic, ...existingTopic}))
+    return Object.values(mergedTopicDict)
+  }), [topicRecommendations])
```

Comment:
This block is pretty hard to review.  The construction of `ownTopicDict` and `matchedPersonTopicDict` can probably be productively broken up into ~3-4 lines (each), maybe inside their own functions.  I'm also not sure why they're being merged.

Unclear why `useEffect` is calling `setTopicPreferences` to assign a different value to `topicPreferences` instead of just setting that value in the initial `useState`.  The logic inside of `useEffect` is also pretty hard to review (too much nesting/inline logic when mapping over stuff).

Reply:
@b0b3rt I think this should be a bit easier to review now